softlayer/resource_softlayer_file_storage.go

```diff
@@ -0,0 +1,941 @@
+			if err != nil {
+				return datatypes.Network_Storage{}, "", err
+			}
+
+			if len(storage) == 1 {
+				return storage[0], "complete", nil
+			} else if len(storage) == 0 {
+				return nil, "pending", nil
+			} else {
+				return nil, "", fmt.Errorf("Expected one Storage: %s", err)
+			}
+		},
+		Timeout:        45 * time.Minute,
+		Delay:          10 * time.Second,
+		MinTimeout:     10 * time.Second,
+		NotFoundChecks: 300,
+	}
+
+	pendingResult, err := stateConf.WaitForState()
+
+	if err != nil {
+		return datatypes.Network_Storage{}, err
+	}
+
+	var result, ok = pendingResult.(datatypes.Network_Storage)
+
+	if ok {
+		return result, nil
+	}
+
+	return datatypes.Network_Storage{},
+		fmt.Errorf("Cannot find Storage with order id '%d'", orderId)
+}
+
+// Waits for storage provisioning
+func WaitForStorageAvailable(d *schema.ResourceData, meta interface{}) (interface{}, error) {
+	log.Printf("Waiting for storage (%s) to be available.", d.Id())
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return nil, fmt.Errorf("The storage ID %s must be numeric", d.Id())
+	}
+	sess := meta.(ProviderConfig).SoftLayerSession()
+
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"retry", "provisioning"},
+		Target:  []string{"available"},
+		Refresh: func() (interface{}, string, error) {
+			// Check active transactions
+			service := services.GetNetworkStorageService(sess)
+			result, err := service.Id(id).Mask("activeTransactions").GetObject()
+			if err != nil {
+				if apiErr, ok := err.(sl.Error); ok && apiErr.StatusCode == 404 {
+					return nil, "", fmt.Errorf("Error retrieving storage: %s", err)
+				}
+				return false, "retry", nil
+			}
+
+			log.Println("Checking active transactions.")
+			if len(result.ActiveTransactions) > 0 {
+				return result, "provisioning", nil
+			}
+
+			// Check volume status.
+			log.Println("Checking volume status.")
+			resultStr := ""
+			err = sess.DoRequest(
+				"SoftLayer_Network_Storage",
+				"getObject",
+				nil,
+				&sl.Options{Id: &id, Mask: "volumeStatus"},
+				&resultStr,
+			)
+			if err != nil {
+				return false, "retry", nil
+			}
+
+			if !strings.Contains(resultStr, "PROVISION_COMPLETED") &&
+				!strings.Contains(resultStr, "Volume Provisioning has completed") {
+				return result, "provisioning", nil
+			}
+
+			return result, "available", nil
+		},
+		Timeout:    45 * time.Minute,
+		Delay:      10 * time.Second,
+		MinTimeout: 10 * time.Second,
+	}
+
+	return stateConf.WaitForState()
+}
+
+func getIopsKeyName(iops float64, storageType string) (string, error) {
+	switch storageType {
+	case EnduranceType:
+		return enduranceIopsMap[iops], nil
+	case PerformanceType:
+		return fmt.Sprintf("%.f_IOPS", iops), nil
+	}
+	return "", fmt.Errorf("Invalid storageType %s.", storageType)
+}
+
+func getPrice(productItems []datatypes.Product_Item, keyName string, categoryCode string, capacityRestrictionType string, capacityRestriction int) (datatypes.Product_Item_Price, error) {
```

Comment:
This function is begging for a unit test. Unit tests here would go a long way towards explaining what this function is capable of, and it would allow developers to extend it without fear of breaking it.